File: src/main/java/com/bajdcc/LALR1/syntax/exp/SequenceExp.java
package com.bajdcc.LALR1.syntax.exp;

import com.bajdcc.LALR1.syntax.ISyntaxComponent;
import com.bajdcc.LALR1.syntax.ISyntaxComponentVisitor;
import com.bajdcc.util.VisitBag;

import java.util.ArrayList;
import java.util.List;

/**
 * 文法规则（序列）
 *
 * @author bajdcc
 */
public class SequenceExp implements ISyntaxComponent, IExpCollction {

	/**
	 * 子表达式表
	 */
	public List<ISyntaxComponent> arrExpressions = new ArrayList<>();

	@Override
	public void visit(ISyntaxComponentVisitor visitor) {
		VisitBag bag = new VisitBag();
		visitor.visitBegin(this, bag);
		if (bag.getVisitChildren()) {
			for (ISyntaxComponent exp : arrExpressions) {
				exp.visit(visitor);
			}
		}
		if (bag.getVisitEnd()) {
			visitor.visitEnd(this);
		}
	}

	@Override
	public void add(ISyntaxComponent exp) {
		arrExpressions.add(exp);
	}

	@Override
	public boolean isEmpty() {
		return arrExpressions.isEmpty();
	}
}

File: src/main/kotlin/com/bajdcc/LL1/syntax/exp/SequenceExp.kt
package com.bajdcc.LL1.syntax.exp

import com.bajdcc.LL1.syntax.ISyntaxComponent
import com.bajdcc.LL1.syntax.ISyntaxComponentVisitor
import com.bajdcc.util.VisitBag

/**
 * 文法规则（序列）
 *
 * @author bajdcc
 */
class SequenceExp : ISyntaxComponent, IExpCollction {

    /**
     * 子表达式表
     */
    var arrExpressions = mutableListOf<ISyntaxComponent>()

    override val isEmpty: Boolean
        get() = arrExpressions.isEmpty()

    override fun visit(visitor: ISyntaxComponentVisitor) {
        val bag = VisitBag()
        visitor.visitBegin(this, bag)
        if (bag.visitChildren) {
            for (exp in arrExpressions) {
                exp.visit(visitor)
            }
        }
        if (bag.visitEnd) {
            visitor.visitEnd(this)
        }
    }

    override fun add(exp: ISyntaxComponent) {
        arrExpressions.add(exp)
    }
}
